### PR TITLE
feat(video-menu): native track volume control updates

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

File: react/features/video-menu/components/native/RemoteVideoMenu.tsx [hunk 1]
```diff
@@ -158,21 +160,33 @@ class RemoteVideoMenu extends PureComponent<IProps> {
                 { !_disablePrivateChat && <PrivateMessageButton { ...buttonProps } /> }
                 <ConnectionStatusButton
                     { ...buttonProps }
+                    /* @ts-ignore */
```
Comment (saghul) [inline react/features/video-menu/components/native/RemoteVideoMenu.tsx:163]: Why is this necessary now?

File: react/features/video-menu/components/native/RemoteVideoMenu.tsx [hunk 2]
```diff
@@ -158,21 +160,33 @@ class RemoteVideoMenu extends PureComponent<IProps> {
...
+                                        key = { room.id }
+                                        /* @ts-ignore */
+                                        room = { room }
+                                        { ...buttonProps } />
+                                )
+                            )
+                        }
+                    </>
+                }
+                {/* @ts-ignore */}
```
Comment (saghul) [inline react/features/video-menu/components/native/RemoteVideoMenu.tsx:188]: Why do we need to ts-ignore here? Can we try to fix it pl?

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 3]
```diff
@@ -23,6 +31,8 @@ interface IProps {
      */
     _startSilent: boolean;
 
+    _track: any;
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:34]: Add a doc comment.

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 4]
```diff
@@ -74,7 +80,7 @@ class VolumeSlider extends PureComponent<IProps, IState> {
         super(props);
 
         this.state = {
-            volumeLevel: props._volume || 0
+            volumeLevel: props._volume || NATIVE_VOLUME_SLIDER_SCALE / 2
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:83]: This breaks if _volume is 0, check it it's undefined instead.

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 5]
```diff
@@ -138,16 +149,18 @@ class VolumeSlider extends PureComponent<IProps, IState> {
...
-function mapStateToProps(state: IReduxState, ownProps: any) {
+function mapStateToProps(state: IReduxState, ownProps: IProps) {
     const { participantID } = ownProps;
-    const { participantsVolume } = state['features/participants-pane'];
+    const { participantsVolume } = state['features/filmstrip'];
     const { startSilent } = state['features/base/config'];
+    const tracks = getTrackState(state);
 
     return {
+        _track: getTrackByMediaTypeAndParticipant(tracks, MEDIA_TYPE.AUDIO, participantID),
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:159]: alpha sort please.

File: react/features/video-menu/constants.ts [hunk 6]
```diff
@@ -4,3 +4,10 @@
  * recognizes whole numbers.
  */
 export const VOLUME_SLIDER_SCALE = 100;
+
+/**
+ * Used to modify initialValue, which is expected to be a decimal value between
+ * 0 and 1, and converts it to a number representable by an input slider, which
+ * recognizes whole numbers.
+ */
+export const NATIVE_VOLUME_SLIDER_SCALE = 10;
```
Comment (saghul) [inline react/features/video-menu/constants.ts:13]: alpha sort please.

File: react/features/video-menu/constants.ts [hunk 7]
```diff
@@ -4,3 +4,10 @@
  * recognizes whole numbers.
  */
 export const VOLUME_SLIDER_SCALE = 100;
+
+/**
+ * Used to modify initialValue, which is expected to be a decimal value between
```
Comment (saghul) [inline react/features/video-menu/constants.ts:9]: Update the comment to match reality.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react/features/video-menu/components/native/RemoteVideoMenu.tsx (hunk 1 above)
Comment (Calinteodor) [reply react/features/video-menu/components/native/RemoteVideoMenu.tsx]: Found a workaround.

File: react/features/video-menu/constants.ts [hunk 8]
```diff
@@ -1,3 +1,8 @@
+/**
+ * Used to set maximumValue for native volume slider.
+ */
+export const NATIVE_VOLUME_SLIDER_SCALE = 20;
```
Comment (saghul) [inline react/features/video-menu/constants.ts:4]: Please document how we use the scale, since it's somewhat unconventional.

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 9]
```diff
@@ -138,16 +161,18 @@ class VolumeSlider extends PureComponent<IProps, IState> {
...
-    const { participantsVolume } = state['features/participants-pane'];
+    const { participantsVolume } = state['features/filmstrip'];
     const { startSilent } = state['features/base/config'];
+    const tracks = getTrackState(state);
 
     return {
         _startSilent: Boolean(startSilent),
-        _volume: participantID && participantsVolume[participantID]
+        _track: getTrackByMediaTypeAndParticipant(tracks, MEDIA_TYPE.AUDIO, participantID),
+        _volume: participantID ? participantsVolume[participantID] : undefined
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:173]: why did you rewrite this prop? the previous form provides the same result.

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 10]
```diff
@@ -123,10 +134,22 @@ class VolumeSlider extends PureComponent<IProps, IState> {
      * @private
      * @returns {void}
      */
-    _onVolumeChange(volumeLevel: number) {
-        const { dispatch, participantID } = this.props;
+    _onVolumeChange(volumeLevel: any) {
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:137]: Why did you change the type to any?

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 11]
```diff
@@ -123,10 +134,22 @@ class VolumeSlider extends PureComponent<IProps, IState> {
...
+        const audioTrack = _track?.jitsiTrack.track;
+
+        let newVolumeLevel;
 
-        dispatch(setVolume(participantID, volumeLevel));
+        if (volumeLevel <= 10) {
+            newVolumeLevel = volumeLevel / 10;
+        } else {
+            newVolumeLevel = volumeLevel;
+        }
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:147]: This doesn't look right. If the new volume is 0-10 we do want it divided by 10. If it's 11-20 we need to put it in the (1-10) range so we need to substract 10.

Now, that would mean that 10 and 11 are the same value, which is arguably wrong, so maybe we should have the scale go from 0 to 19, so 20 elements total.

0-10: value / 10
11-19: value -9

File: react/features/video-menu/components/native/VolumeSlider.tsx [hunk 12]
```diff
@@ -74,7 +83,7 @@ class VolumeSlider extends PureComponent<IProps, IState> {
         super(props);
 
         this.state = {
-            volumeLevel: props._volume || 0
+            volumeLevel: props._volume ?? NATIVE_VOLUME_SLIDER_SCALE / 2
```
Comment (saghul) [inline react/features/video-menu/components/native/VolumeSlider.tsx:86]: Note that this won't work when you change the scale to 0-19.

**Review by saghul — DISMISSED** [review]

File: react/features/video-menu/components/native/VolumeSlider.tsx (hunk 10 above)
Comment (Calinteodor) [reply react/features/video-menu/components/native/VolumeSlider.tsx:137]: TS error.

File: react/features/video-menu/components/native/VolumeSlider.tsx (hunk 9 above)
Comment (Calinteodor) [reply react/features/video-menu/components/native/VolumeSlider.tsx:173]: Reverted.

**Review by saghul — APPROVED** [review]
🚀
